DisCatSharp.ApplicationCommands/ApplicationCommandsExtension.cs

```diff
@@ -2141,4 +2150,189 @@ internal class ContextMenuCommand
+                    return options.AsEnumerable();
+                }
+                var foundCommand = command.Options.ToList().FindAll(op => op.Name.ToLower().Equals(context.Options[1].Value.ToString().ToLower().Trim())).FirstOrDefault();
+                if (foundCommand is null || foundCommand.Options is null)
+                {
+                    options.Add(new DiscordApplicationCommandAutocompleteChoice("no_options_for_this_command", "no_options_for_this_command"));
+                }
+                else
+                {
+                    foundCommand.Options.ToList().ForEach(option => options.Add(new DiscordApplicationCommandAutocompleteChoice(option.Name, option.Name)));
+                }
+                return options.AsEnumerable();
+            }
+        }
+
+        [SlashCommand("help", "Displays command help")]
+        public async Task DefaultHelpAsync(InteractionContext ctx,
+            [Autocomplete(typeof(DefaultHelpAutoCompleteProvider))]
+            [Option("option_one", "top level command to provide help for", true)] string commandName,
+            [Autocomplete(typeof(DefaultHelpAutoCompleteLevelOneProvider))]
+            [Option("option_two", "subgroup or command to provide help for", true)] string commandOneName = null,
+            [Autocomplete(typeof(DefaultHelpAutoCompleteLevelTwoProvider))]
+            [Option("option_three", "command to provide help for", true)] string commandTwoName = null)
+        {
+            ulong currentGuildId = ctx.Guild.Id;
+            var applicationCommands = new List<DiscordApplicationCommand>();
+            var globalCommands = await ctx.Client.GetGlobalApplicationCommandsAsync();
+            var guildCommands = await ctx.Client.GetGuildApplicationCommandsAsync(ctx.Guild.Id);
+            applicationCommands.AddRange(globalCommands);
+            applicationCommands.AddRange(guildCommands);
+            applicationCommands.Distinct().ToList();
+            applicationCommands.RemoveAll(ac => ac.Name.ToLower().Equals("help"));
+            if (applicationCommands is null)
+            {
+                await ctx.CreateResponseAsync(InteractionResponseType.ChannelMessageWithSource, new DiscordInteractionResponseBuilder()
+                    .WithContent($"There are no slash commands for guild {ctx.Guild.Name}").AsEphemeral(true));
+                return;
+            }
+            if (commandTwoName is not null && !commandTwoName.Equals("no_options_for_this_command"))
+            {
+                List<DiscordApplicationCommand> commandsWithSubCommands = applicationCommands.FindAll(ac => ac.Options is not null && ac.Options.Any(op => op.Type == ApplicationCommandOptionType.SubCommandGroup));
+                var cmdParent = commandsWithSubCommands.FirstOrDefault(cm => cm.Options.ToList().Any(op => op.Name.Equals(commandOneName))).Options.ToList()
```

Comment:
Converting these sources into Lists before using other linq methods is not needed. So long as the source is IEnumerable / collection of sorts you can use a linq method.

Now, at the very very end if you want the end result to be a list -- yes you should call `ToList`